UbuntuDrivers/detect.py

```diff
@@ -70,58 +70,91 @@ def system_modaliases(sys_path=None):
     return aliases
 
 
-def _check_video_abi_compat(apt_cache, record):
+def _check_video_abi_compat(apt_cache, package):
     xorg_video_abi = None
 
-    # determine current X.org video driver ABI
-    try:
-        for p in apt_cache['xserver-xorg-core'].candidate.provides:
-            if p.startswith('xorg-video-abi-'):
-                xorg_video_abi = p
-                # logging.debug('_check_video_abi_compat(): Current X.org video abi: %s', xorg_video_abi)
-                break
-    except (AttributeError, KeyError):
-        logging.debug('_check_video_abi_compat(): xserver-xorg-core not available, cannot check ABI')
+    if package.name.startswith('nvidia-driver-'):
+        xorg_driver_name = package.name.replace('nvidia-driver-', 'xserver-xorg-video-nvidia-')
+        try:
+            package = apt_cache.__getitem__(xorg_driver_name)
+        except KeyError:
+            logging.debug('Cannot find %s package in the cache. Cannot check ABI' % (xorg_driver_name))
+            return True
+
+    records = apt_pkg.PackageRecords(apt_cache)
+    depcache = apt_pkg.DepCache(apt_cache)
+    candidate = depcache.get_candidate_ver(package)
+    records.lookup(candidate.file_list[0])
+    deps = records.__getitem__('Depends').split(', ')
+    needs_video_abi = False
+    for dep in deps:
+        if dep.startswith('xorg-video-abi-'):
+            needs_video_abi = True
+            break
+
+    if not needs_video_abi:
+        logging.debug('Skipping check for %s since it does not depend on video abi' % package.name)
         return True
-    if not xorg_video_abi:
-        return False
 
+    # determine current X.org video driver ABI
     try:
-        deps = record['Depends']
+        xorg_core = apt_cache.__getitem__('xserver-xorg-core')
     except KeyError:
+        logging.debug('xserver-xorg-core not available, cannot check ABI')
         return True
-    if 'xorg-video-abi-' in deps and xorg_video_abi not in deps:
+
+    candidate = depcache.get_candidate_ver(xorg_core)
+    records.lookup(candidate.file_list[0])
+
+    provides = records.__getitem__('Provides').split(', ')
```

Comment:
No need for records, Use candidate.provides_list instead, that's more correct and much faster.